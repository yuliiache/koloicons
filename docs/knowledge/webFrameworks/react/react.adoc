== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







