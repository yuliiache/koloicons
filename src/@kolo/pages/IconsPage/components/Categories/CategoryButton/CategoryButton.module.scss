@import "src/@kolo/design";

.categoryButton {
  background-color: transparent;
  width: 100%;
  min-height: 32px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 16px;
  color: $secFontColor;
  border-radius: 4px;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-size: 14px;
  line-height: 1.43;
  text-align: left;
  border: 0;
  cursor: pointer;

  &.categoryButton-active {
    border-radius: 4px;
    font-weight: 500;
    background-color: $BlueColor;
    color: $WhiteBG;
    outline: none;
  }
}
