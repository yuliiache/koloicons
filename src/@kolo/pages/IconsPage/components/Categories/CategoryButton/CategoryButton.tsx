import classnames from 'classnames';
import React, {MouseEventHandler} from 'react';

import styles from './CategoryButton.module.scss';

interface CategoryButtonProps {
  name: string;
  onClick: MouseEventHandler<HTMLButtonElement>;
  isActive: boolean;
}

const CategoryButton: React.FC<CategoryButtonProps> = ({name, onClick, isActive}) => {
  const buttonClasses = classnames(styles.categoryButton, {
    [styles['categoryButton-active']]: isActive,
  });

  return (
    <button
      className={buttonClasses}
      onClick={onClick}
    >
      {name}
    </button>
  );
};

export default CategoryButton;
