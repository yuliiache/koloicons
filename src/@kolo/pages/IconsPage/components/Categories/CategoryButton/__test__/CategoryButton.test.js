import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';

import '@testing-library/jest-dom';

import CategoryButton from '../CategoryButton';

describe('CategoryButton', () => {
  it('renders correctly with given name and matches snapshot', () => {
    const name = 'Test Category';
    const {asFragment} = render(
      <CategoryButton
        name={name}
        onClick={() => {}}
        isActive={false}
      />
    );
    expect(screen.getByRole('button')).toHaveTextContent(name);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onClick prop when clicked', () => {
    const handleClick = jest.fn();
    render(
      <CategoryButton
        name="Test"
        onClick={handleClick}
        isActive={false}
      />
    );
    fireEvent.click(screen.getByRole('button'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('applies active class when isActive is true', () => {
    const {container} = render(
      <CategoryButton
        name="Test"
        onClick={() => {}}
        isActive={true}
      />
    );
    expect(container.firstChild).toHaveClass('categoryButton-active');
  });

  it('does not apply active class when isActive is false', () => {
    const {container} = render(
      <CategoryButton
        name="Test"
        onClick={() => {}}
        isActive={false}
      />
    );
    expect(container.firstChild).not.toHaveClass('categoryButton-active');
  });
});
