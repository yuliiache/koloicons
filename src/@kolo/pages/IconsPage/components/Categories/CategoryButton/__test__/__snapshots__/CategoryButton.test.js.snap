// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryButton renders correctly with given name and matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="categoryButton"
  >
    Test Category
  </button>
</DocumentFragment>
`;
